This README describes the scripts used for variant calling with GATK in the DD cohort

# Scripts
This lists all of the scripts in the directory and subdirectories and the original scripts each file was derived from and a description
- _1_GATK_pipeline.sh_: Run BWA alignment and GATK on FASTQ files to call CNV variants
- _2_genotypeGVCF.sh_: Perform joint genotyping across the cohort using GATK
- _3_split_norm.sh_: Split the cohort pVCF and normalize variants

